lib/Dialect/FIRRTL/FIRRTLReductions.cpp

```diff
@@ -1134,7 +1134,15 @@ struct EagerInliner : public OpReduction<InstanceOp> {
   void beforeReduction(mlir::ModuleOp op) override {
     symbols.clear();
     nlaRemover.clear();
-    nlaTable = std::make_unique<NLATable>(op);
+    CircuitOp circuit;
+    for (auto &topLevelOp : *op.getBody())
+      if ((circuit = dyn_cast<CircuitOp>(topLevelOp)))
+        break;
+
+    if (circuit)
+      nlaTable = std::make_unique<NLATable>(circuit.getOperation());
+    else
+      nlaTable.reset();
```

Comment:
I wonder if we could create a dense map of multiple NLATables, one for each circuit. That would allow us to run this reduction on inputs with multiple circuit ops. Maybe something like this works:
```c++
nlaTables.clear();
for (auto circuit : op.getOps<CircuitOp>())
  nlaTables.insert({circuit, std::make_unique<NLATable>(circuit)});
```
We could then define the tables using a map instead of a single unique pointer:
```c++
DenseMap<CircuitOp, std::unique_ptr<NLATable>> nlaTables;
```
And inside the reduction when we need the NLATable, we can look it up using the local circuit:
```c++
auto &nlaTable = nlaTables.lookup(op.getParentOfType<CircuitOp>());
if (!nlaTable)
  return 0;
```
What do you think about this?

Reply:
Sounds good! I'll implement your suggestions and test it on my local setup :)